src/placeos-driver/interface/sensor.cr

```diff
@@ -0,0 +1,220 @@
+    SensorType::Speed                => Unit::MetrePerSecond,
+    SensorType::Roll                 => Unit::Angle,
+    SensorType::Pitch                => Unit::Angle,
+    SensorType::Yaw                  => Unit::Angle,
+    SensorType::Compass              => Unit::Angle,
+    SensorType::Current              => Unit::Ampere,
+    SensorType::Voltage              => Unit::Volt,
+    SensorType::ElectricalResistance => Unit::Ohm,
+    SensorType::Power                => Unit::Watt,
+    SensorType::Radiation            => Unit::Sievert,
+    SensorType::Distance             => Unit::Metre,
+    SensorType::Area                 => Unit::SquareMeter,
+    SensorType::SoundPressure        => Unit::Decibel,
+    SensorType::Capacitance          => Unit::Farad,
+    SensorType::Inductance           => Unit::Henry,
+    SensorType::Conductance          => Unit::Siemens,
+    SensorType::MagneticFlux         => Unit::Weber,
+    SensorType::MagneticFluxDensity  => Unit::Tesla,
+    SensorType::Energy               => Unit::WattSecond,
+    SensorType::Force                => Unit::Newton,
+    SensorType::Frequency            => Unit::Hertz,
+    SensorType::Mass                 => Unit::Kilogram,
+    SensorType::Momentum             => Unit::NewtonSecond,
+    SensorType::TimePeriod           => Unit::Second,
+    SensorType::Volume               => Unit::Litre,
+    SensorType::Acidity              => Unit::PH,
+  }
+
+  enum Status
+    Normal
+    Alarm
+    Fault
+    OutOfService
+  end
+
+  # return the specified sensor details
+  abstract def sensor(unique_id : String) : Detail?
+
+  # return an array of sensor details
+  # zone_id can be ignored if location is unknown by the sensor provider
+  # mac_address can be used to grab data from a single device (basic grouping)
+  abstract def sensors(type : String? = nil, mac_address : String? = nil, zone_id : String? = nil) : Array(Detail)
+
+  abstract class Location
+    include JSON::Serializable
+
+    use_json_discriminator "type", {
+      "geo" => GeoLocation,
+      "map" => MapLocation,
+    }
+  end
+
+  class GeoLocation < Location
+    getter type : String = "geo"
+
+    def initialize(@lat, @lon)
+    end
+
+    property lat : Float64
+    property lon : Float64
+
+    def s2_cell_id(s2_level : Int32 = 21)
+      S2Cells::LatLon.new(lat, lon).to_token(s2_level)
+    end
+  end
+
+  class MapLocation < Location
+    getter type : String = "map"
+
+    def initialize(@x, @y)
+    end
+
+    # this is expected to be the raw unadjusted values
+    property x : Float64
+    property y : Float64
+  end
+
+  abstract class Detail
+    include JSON::Serializable
+    include JSON::Serializable::Unmapped
+
+    def initialize(
+      @type, @value, @unix_ms, @mac, @id,
+      @name, @raw, @loc, @status = Status::Normal
+    )
+    end
+
+    property status : Status
+    property type : SensorType
+
+    @[JSON::Field(ignore: true)]
+    getter unit : Unit { SENSOR_UNIT[@type] }
```

Comment:
```suggestion
    delegate unit, to: type
```

Since `type` is a property, I don't think `unit` should be a getter as `type` can change.